### PR TITLE
Refactor runtime crates into feature-guarded modules

### PR DESCRIPTION
**Pull request by jcrossley3**
To address #130

### CODE REVIEW

**Comment by slinkydeveloper** [comment]
Ok the first thing I would do is to split this PR into several PRs, one per submodule :)

**Comment by slinkydeveloper** [comment]
You also need to sign the DCO :)

File: src/lib.rs [hunk 1]
```diff
@@ -40,6 +40,15 @@
 #![doc(html_root_url = "https://docs.rs/cloudevents-sdk/0.3.1")]
 #![deny(broken_intra_doc_links)]
 
+#[cfg(feature = "ce-actix")]
```
Comment (slinkydeveloper) [inline src/lib.rs:43]: Just name the feature `actix` without prepending with ce? Same for the below modules

File: src/actix/server_request.rs [hunk 2]
```diff
@@ -134,6 +135,17 @@ impl HttpRequestExt for HttpRequest {
...
+    type Config = ();
+    type Error = actix_web::Error;
+    type Future = Ready<std::result::Result<Event, actix_web::Error>>;
+
+    fn from_request(req: &HttpRequest, payload: &mut actix_web::dev::Payload) -> Self::Future {
+        // TODO: request_to_event(req, web::Payload(payload.take()))
+        ok(Event::default())
+    }
+}
+
```
Comment (slinkydeveloper) [inline src/actix/server_request.rs:148]: Can we do this in another PR, after this PR to move the code?

**Comment by jcrossley3** [comment]
> You also need to sign the DCO :)

I signed all but the one I need help with. I can do that when I figure out how to fix it.

File: src/lib.rs (hunk 1 above)
Comment (jcrossley3) [reply src/lib.rs:43]: That was my initial approach. As you can see from my [2nd commit](https://github.com/cloudevents/sdk-rust/pull/131/commits/14ac80ebfb09bd5cff527d0404ba3b5551a2029d), I had a conflict with `reqwest`. I'm happy to resolve that however you like, but I opted for a consistent prefix.

File: src/actix/server_request.rs (hunk 2 above)
Comment (jcrossley3) [reply src/actix/server_request.rs:148]: Sure thing. I can back it out.

**Comment by jcrossley3** [comment]
> Ok the first thing I would do is to split this PR into several PRs, one per submodule. Also don't remove the old packages now, we can do that later.

So you prefer 1-PR-per-submodule to 1-commit-per-submodule as I currently have it? I can do that, but I'm curious why that's better? Once I back out the deletion of the existing packages and `FromRequest` impl, we'll only be left with the 4 submodule commits in this PR. Is that better than 4 PR's with one commit each?

**Comment by jcrossley3** [comment]
Ok, I just force-pushed changes that leave the old package dirs and don't add the `FromRequest` impl. Just one commit for each submodule feature. I left the `ce-` prefix in for now, and I tried to address the test failures as well.

File: src/actix/headers.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
+use crate::event::SpecVersion;
+use actix_web::http::header;
+use actix_web::http::{HeaderName, HeaderValue};
+use lazy_static::lazy_static;
+use std::collections::HashMap;
+use std::str::FromStr;
+
+macro_rules! unwrap_optional_header {
```
Comment (slinkydeveloper) [inline src/actix/headers.rs:8]: Not for this PR, but maybe for future refactorings: a lot of these macros are copy paste, because the APIs are practically the same even if their names are different, so it would be nice to find a way to unify them

File: src/actix/mod.rs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+//!                 .data("application/json", json!({"hello": "world"}))
+//!                 .build()
+//!                 .expect("No error while building the event"),
+//!         )
+//!         .await?
+//!     )
+//! }
+//! ```
+//!
+//! Check out the [cloudevents-sdk](https://docs.rs/cloudevents-sdk) docs for more details on how to use [`cloudevents::Event`]
```
Comment (slinkydeveloper) [inline src/actix/mod.rs:42]: This is not useful anymore

Comment (slinkydeveloper) [reply src/actix/mod.rs:42]: Same comment applies to other commits as well

File: src/actix/mod.rs [hunk 5]
```diff
@@ -0,0 +1,57 @@
+//! This crate integrates the [cloudevents-sdk](https://docs.rs/cloudevents-sdk) with [Actix web](https://docs.rs/actix-web/) to easily send and receive CloudEvents.
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply src/actix/mod.rs:1]: Same comment applies to other commits as well

File: src/actix/mod.rs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+//!         )
+//!         .await?
+//!     )
+//! }
+//! ```
+//!
+//! Check out the [cloudevents-sdk](https://docs.rs/cloudevents-sdk) docs for more details on how to use [`cloudevents::Event`]
+
+#![doc(html_root_url = "https://docs.rs/cloudevents-sdk/0.3.1")]
+#![deny(broken_intra_doc_links)]
```
Comment (slinkydeveloper) [inline src/actix/mod.rs:45]: I think these are not useful anymore

Comment (slinkydeveloper) [reply src/actix/mod.rs:45]: Same comment applies to other commits as well

**Review by slinkydeveloper — COMMENTED** [review]
Can you try to run locally `cargo doc --open --all-features` to see if this breaks the doc build? You should also try `cargo publish --dry-run` to see if any publish check breaks (although I'm not sure you can do that without logging in)

**Comment by slinkydeveloper** [comment]
Another thing I'm not 100% sure of is how this is gonna play out with our existing CI config, with our matrix of different targets... For example, the actix feature is not supposed to build with wasm

File: cloudevents-sdk-actix-web/Cargo.toml [hunk 7]
```diff
@@ -13,11 +13,11 @@ categories = ["web-programming", "encoding", "web-programming::http-server"]
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-cloudevents-sdk = { version = "0.3.0", path = ".." }
+cloudevents-sdk = { path = ".." }
```
Comment (slinkydeveloper) [inline cloudevents-sdk-actix-web/Cargo.toml:16]: You need to keep the version, it's required when publishing

Comment (jcrossley3) [reply cloudevents-sdk-actix-web/Cargo.toml:16]: Does this matter if we're going to delete the workspaces? Or do you want to continue to publish all of them even though the logic is replicated in the root? The `cargo publish --dry-run` from the root worked fine.

**Comment by jcrossley3** [comment]
> Can you try to run locally `cargo doc --open --all-features` to see if this breaks the doc build? You should also try `cargo publish --dry-run` to see if any publish check breaks (although I'm not sure you can do that without logging in)

Done. Fixed several broken links. I only tested the publish from the root and it seemed fine, abandoning only because of `--dry-run` and without requiring me to log in.

**Comment by jcrossley3** [comment]
> Another thing I'm not 100% sure of is how this is gonna play out with our existing CI config, with our matrix of different targets... For example, the actix feature is not supposed to build with wasm. I might be wrong, but it seems to me that the integrations are not tested anymore with this PR, without modifying the existing github actions config.

I added an `--all-features` to every step that had `--workspace`. This essentially runs all the tests twice. When we delete the workspace dirs, we can remove the `--workspace` option in the actions, too.

As far as the wasm32 build goes, I added the `cloudevents-reqwest` feature to the args, but I'm unable to see this working locally. I can't get past the following:

```console
[jim@localhost reqwest-wasm-example]$ npm run serve

> @ serve /home/jim/src/rust/sdk-rust/example-projects/reqwest-wasm-example
> webpack-dev-server

🧐  Checking for wasm-pack...

ℹ️  Installing wasm-pack 

ℹ ｢wds｣: Project is running at http://localhost:8080/
ℹ ｢wds｣: webpack output is served from /
ℹ ｢wds｣: Content not from webpack is served from /home/jim/src/rust/sdk-rust/example-projects/reqwest-wasm-example
✖ ｢wdm｣: Error: Rust compilation.
    at ChildProcess.<anonymous> (/home/jim/src/rust/sdk-rust/example-projects/reqwest-wasm-example/node_modules/@wasm-tool/wasm-pack-plugin/plugin.js:210:16)
    at ChildProcess.emit (events.js:376:20)
    at maybeClose (internal/child_process.js:1055:16)
    at Socket.<anonymous> (internal/child_process.js:441:11)
    at Socket.emit (events.js:376:20)
    at Pipe.<anonymous> (net.js:673:12)
```

File: cloudevents-sdk-actix-web/Cargo.toml (hunk 7 above)
Comment (slinkydeveloper) [reply cloudevents-sdk-actix-web/Cargo.toml:16]: > Or do you want to continue to publish all of them even though the logic is replicated in the root?

Just for another version, then we'll remove them

> https://github.com/vert-x3/vertx-eventbus-bridge-clients/tree/master/Go

you need to try `cargo publish --dry-run` from the subdirs as well

**Comment by slinkydeveloper** [comment]
> As far as the wasm32 build goes, I added the cloudevents-reqwest feature to the args, but I'm unable to see this working locally. I can't get past the following:

Mhhh that error is weird, perhaps the js deps needs to be updated

File: cloudevents-sdk-actix-web/Cargo.toml (hunk 7 above)
Comment (jcrossley3) [reply cloudevents-sdk-actix-web/Cargo.toml]: Done. I didn't touch `warp` though, since it already lacked a `version`. Never published, I guess?

**Comment by jcrossley3** [comment]
> > As far as the wasm32 build goes, I added the cloudevents-reqwest feature to the args, but I'm unable to see this working locally. I can't get past the following:
> 
> Mhhh that error is weird, perhaps the js deps needs to be updated

Didn't help. I get the same error on `master`, so I don't think this PR broke it. Besides, all the checks are passing. ;)

I think I've addressed all your concerns. Let me know if you need anything else!

File: src/actix/headers.rs (hunk 3 above)
Comment (jcrossley3) [reply src/actix/headers.rs:8]: Addressed in #146
